opal/util/json/opal_json.c

```diff
@@ -0,0 +1,263 @@
+        *out = OPAL_JSON_TYPE_COUNT;
+        ret = OPAL_ERROR;
+    }
+
+    return ret;
+}
+
+static int opal_json_internal_new(const json_value *in, opal_json_internal_t **out)
+{
+    *out = malloc(sizeof(opal_json_internal_t));
+    if (NULL == *out) {
+        return OPAL_ERROR;
+    }
+
+    (*out)->value = (json_value *) in;
+    return opal_json_internal_translate_type(in->type, &(*out)->parent.type);
+}
+
+int opal_json_load(const char *str, const size_t len, opal_json_t **json)
+{
+    opal_json_internal_t *out = NULL;
+    int ret = OPAL_SUCCESS;
+
+    json_value *value = json_parse(str, len);
+    if (!value) {
+        opal_output_verbose(1, 0, "Unable to parse JSON string\n");
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    ret = opal_json_internal_new(value, &out);
+
+out:
+    if (OPAL_SUCCESS == ret) {
+        *json = (opal_json_t *) out;
+    } else if (out) {
+        opal_json_internal_free(out);
+    }
+
+    return ret;
+}
+
+int opal_json_load_file(const char *filename, opal_json_t **json)
+{
+    FILE *fp = NULL;
+    size_t file_size;
+    char *file_contents = NULL;
+    int ret = OPAL_SUCCESS;
+
+    fp = fopen(filename, "r");
+    if (fp == NULL) {
+        opal_output_verbose(1, 0, "Unable to open %s\n", filename);
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    (void) fseek(fp, 0L, SEEK_END);
+    file_size = ftell(fp);
+    rewind(fp);
+
+    file_contents = (char *) malloc(file_size);
+    if (!file_contents) {
+        opal_output_verbose(1, 0, "Unable to allocate memory\n");
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    if (file_size > fread(file_contents, 1, file_size, fp)) {
+        opal_output_verbose(1, 0, "Unable to read content of %s\n", filename);
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    ret = opal_json_load(file_contents, file_size, json);
+
+out:
+    if (fp) {
+        fclose(fp);
+    }
+    if (file_contents) {
+        free(file_contents);
+    }
+
+    return ret;
+}
+
+int opal_json_get_key(const opal_json_t *json, const char *key, opal_json_t **out)
+{
+    int ret = OPAL_ERROR;
+
+    CHECK_OBJ_TYPE(json, OPAL_JSON_OBJECT);
+
+    opal_json_internal_t *in = (opal_json_internal_t *) json;
+    opal_json_internal_t *result = NULL;
+
+    json_object_entry entry = {0};
+
+    for (unsigned int i = 0; i < in->value->u.object.length; ++i) {
+        entry = in->value->u.object.values[i];
+        if (0 == strcmp(entry.name, key)) {
+            ret = opal_json_internal_new(entry.value, &result);
+            break;
+        }
+    }
+
+    if (OPAL_SUCCESS == ret) {
+        *out = (opal_json_t *) result;
+    } else if (result) {
+        opal_json_internal_free(result);
+    }
+
+    return ret;
+}
+
+void opal_json_free(opal_json_t *json)
+{
+    opal_json_internal_free((struct opal_json_internal_t *) json);
+}
+
+int opal_json_get_index(const opal_json_t *json, const size_t index, opal_json_t **out)
```

Comment:
similar to the comment above, all output opal_json_t should be `const opal_json_t **`.